src/java/htsjdk/samtools/SRAIterator.java

```diff
@@ -0,0 +1,195 @@
+import ngs.ErrorMsg;
+import ngs.ReadCollection;
+import ngs.Reference;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class SRAIterator implements SAMRecordIterator {
+    private ValidationStringency validationStringency;
+
+    private ReadCollection run;
+    private SAMFileHeader header;
+    private final List<Reference> cachedReferences;
+    private RecordRangleInfo recordRangleInfo;
+    private Iterator<Chunk> chunksIterator;
+    private Chunk currentChunk;
+
+    private SRAAlignmentIterator alignmentIterator;
+    private SRAUnalignmentIterator unalignmentIterator;
+
+    public static class RecordRangleInfo {
+        private List<Long> referenceOffsets;
+        private List<Long> referenceLengths;
+        private long totalReferencesLength;
+        private long numberOfReads; // is used for unaligned read space
+        private long totalRecordRangeLength;
+
+        public RecordRangleInfo(List<Long> referenceLengths, long numberOfReads) {
+            this.numberOfReads = numberOfReads;
+            this.referenceLengths = referenceLengths;
+
+            referenceOffsets = new ArrayList<Long>();
+
+            totalReferencesLength = 0;
+            for (Long refLen : referenceLengths) {
+                referenceOffsets.add(totalReferencesLength);
+                totalReferencesLength += refLen;
+            }
+
+            totalRecordRangeLength = totalReferencesLength + this.numberOfReads;
+        }
+
+        public long getNumberOfReads() {
+            return numberOfReads;
+        }
+
+        public long getTotalReferencesLength() {
+            return totalReferencesLength;
+        }
+
+        public long getTotalRecordRangeLength() {
+            return totalRecordRangeLength;
+        }
+
+        public final List<Long> getReferenceOffsets() {
+            return referenceOffsets;
+        }
+
+        public final List<Long> getReferenceLengths() {
+            return referenceLengths;
+        }
+    }
+
+    public static RecordRangleInfo getRecordsRangeInfo(ReadCollection run) {
+        try {
+            return new RecordRangleInfo(SRAAlignmentIterator.getReferencesLengths(run), SRAUnalignmentIterator.getNumberOfReads(run));
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public SRAIterator(final ReadCollection run, final SAMFileHeader header, final List<Reference> cachedReferences,
+                       final RecordRangleInfo recordRangleInfo, final List<Chunk> chunks) {
+        this.run = run;
+        this.header = header;
+        this.cachedReferences = cachedReferences;
+        this.recordRangleInfo = recordRangleInfo;
+        chunksIterator = chunks.iterator();
+        if (chunksIterator.hasNext()) {
+            currentChunk = chunksIterator.next();
+        }
+
+        hasNext();
+    }
+
+    @Override
+    public boolean hasNext() {
```

Comment:
Unfortunately it is not possible because of iterators model that is used in NGS.

In NGS an iterator only has a method
```
boolean nextAlignment() throws ErrorMsg;
```
or
```
boolean nextRead() throws ErrorMsg;
```
which actually moves the iterator to a next position and returns true if it succeeded, or false when there are no more entries. So, to check if there is a next object we have to move internal iterator forward. But this will invalidate previous object contained within the iterator. And since we use SRALazyRecord which has a reference to the iterator and only loads fields from the internal iterator upon request, we will have to immediately break a link from SRALazyRecord to the iterator and load all the fields upfront (even if nobody ever tries to access them). This way we will be able access to "hasNext" value.